compiler/rustc_data_structures/src/xxhash/mod.rs

```diff
@@ -0,0 +1,1204 @@
+            let data_key_lo = _mm_shuffle_epi32(data_key, _mm_shuffle(0, 3, 0, 1));
+            // product     = (data_key & 0xffffffff) * (data_key_lo & 0xffffffff);
+            let product = _mm_mul_epu32(data_key, data_key_lo);
+            // acc[i] += swap(data_vec);
+            let data_swap = _mm_shuffle_epi32(data_vec, _mm_shuffle(1, 0, 3, 2));
+            let sum = _mm_add_epi64(acc.offset(i).read(), data_swap);
+            // acc[i] += product;
+            acc.offset(i).write(_mm_add_epi64(product, sum));
+        }
+    }
+}
+
+#[inline]
+fn XXH_readLE64(ptr: Ptr<u8>) -> u64 {
+    #[cfg(debug_assertions)]
+    ptr.assert_contains_next_n_bytes(8);
+
+    let ptr = ptr.raw() as *const [u8; 8];
+    let bytes = unsafe { *ptr };
+    u64::from_le_bytes(bytes)
+}
+
+#[inline]
+fn XXH_readLE32(ptr: Ptr<u8>) -> u32 {
+    #[cfg(debug_assertions)]
+    ptr.assert_contains_next_n_bytes(4);
+
+    let ptr = ptr.raw() as *const [u8; 4];
+    let bytes = unsafe { *ptr };
+    u32::from_le_bytes(bytes)
+}
+
+#[inline]
+fn XXH_mult32to64(x: u64, y: u64) -> u64 {
+    (x as u32 as u64).wrapping_mul(y as u32 as u64)
+}
+
+#[inline]
+fn XXH3_consumeStripes(
+    acc: &mut Acc,
+    num_stripes_so_far: &mut usize,
+    num_stripes_per_block: usize,
+    input: Ptr<u8>,
+    num_stripes: usize,
+    secret: Ptr<u8>,
+    secret_limit: usize,
+    f_acc512: XXH3_f_accumulate_512,
+    f_scramble: XXH3_f_scrambleAcc,
+) {
+    debug_assert!(num_stripes <= num_stripes_per_block);
+    debug_assert!(*num_stripes_so_far < num_stripes_per_block);
+    if num_stripes_per_block - *num_stripes_so_far <= num_stripes {
+        /* need a scrambling operation */
+        let num_stripes_to_end_of_block = num_stripes_per_block - *num_stripes_so_far;
+        let num_stripes_after_block = num_stripes - num_stripes_to_end_of_block;
+        XXH3_accumulate(
+            acc,
+            input,
+            secret.offset(*num_stripes_so_far * XXH_SECRET_CONSUME_RATE),
+            num_stripes_to_end_of_block,
+            f_acc512,
+        );
+        f_scramble(acc, secret.offset(secret_limit));
+        XXH3_accumulate(
+            acc,
+            input.offset(num_stripes_to_end_of_block * XXH_STRIPE_LEN),
+            secret,
+            num_stripes_after_block,
+            f_acc512,
+        );
+        *num_stripes_so_far = num_stripes_after_block;
+    } else {
+        XXH3_accumulate(
+            acc,
+            input,
+            secret.offset(*num_stripes_so_far * XXH_SECRET_CONSUME_RATE),
+            num_stripes,
+            f_acc512,
+        );
+        *num_stripes_so_far += num_stripes;
+    }
+}
+
+#[inline]
+fn XXH3_accumulate(
+    acc: &mut Acc,
+    input: Ptr<u8>,
+    secret: Ptr<u8>,
+    num_stripes: usize,
+    f_acc512: XXH3_f_accumulate_512,
+) {
+    for n in 0..num_stripes {
+        let input = input.offset(n * XXH_STRIPE_LEN);
+
+        // This is taken from the C implementation. Not sure if it is completely safe.
```

Comment:
I *think* that prefetching an invalid address (in case the implementation reaches out of bounds) should be safe and doesn't affect program behavior. That being said, I would leave microoptimizations like prefetching for later, I will remove this for now.